File: fetch-posts.py
#!/usr/bin/env python3

import requests
import os
import re
from datetime import datetime
from urllib.parse import urlparse
import json
import html2text

def sanitize_filename(title):
    """Convert a title into a valid filename."""
    # Remove any non-alphanumeric characters (except spaces and hyphens)
    filename = re.sub(r'[^\w\s-]', '', title)
    # Replace spaces with hyphens
    filename = re.sub(r'[-\s]+', '-', filename)
    # Convert to lowercase
    return filename.lower()

def get_date_from_post(post):
    """Extract date from post and format it as YYYY-MM-DD."""
    # Parse the date string from the post
    post_date = datetime.strptime(post['date'], '%Y-%m-%dT%H:%M:%S')
    return post_date.strftime('%Y-%m-%d')

def fetch_and_save_posts():
    # WordPress REST API endpoint
    base_url = 'https://developer.woocommerce.com/wp-json/wp/v2'
    
    # First, get the category ID for 'Release Posts'
    categories_url = f'{base_url}/categories'
    categories_response = requests.get(categories_url)
    categories = categories_response.json()
    
    # Find the Release Posts category
    release_posts_category = next(
        (cat for cat in categories if cat['name'] == 'Release Posts'),
        None
    )
    
    if not release_posts_category:
        print("Could not find 'Release Posts' category")
        return
    
    # Get posts from the Release Posts category
    posts_url = f'{base_url}/posts'
    params = {
        'categories': release_posts_category['id'],
        'per_page': 100,  # Adjust this number as needed
        'orderby': 'date',
        'order': 'desc'
    }
    
    posts_response = requests.get(posts_url, params=params)
    posts = posts_response.json()
    
    # Create release-posts directory if it doesn't exist
    output_dir = 'release-posts'
    os.makedirs(output_dir, exist_ok=True)
    
    # Get list of existing files to avoid duplicates
    existing_files = set(os.listdir(output_dir))
    
    # Process each post
    for post in posts:
        # Get the date and title
        post_date = get_date_from_post(post)
        title = post['title']['rendered']
        
        # Skip posts with specific terms in the title
        skip_terms = ['woocommerce-blocks', 'delayed', 'dot-release']
        if any(term.lower() in sanitize_filename(title).lower() for term in skip_terms):
            print(f"Skipping filtered post: {title}")
            continue
        
        # Create a filename with date prefix
        filename = f"{post_date}-{sanitize_filename(title)}.txt"
        
        # Skip if we've already processed this post
        if filename in existing_files:
            print(f"Skipping already processed post: {title}")
            continue
        
        # Prepare the content
        content = f"Title: {title}\n"
        content += f"Date: {post['date']}\n"
        content += f"Link: {post['link']}\n"
        content += "\nContent:\n"
        # Convert HTML to Markdown
        markdown_content = html2text.html2text(post['content']['rendered'])
        content += markdown_content
        
        # Save to file
        filepath = os.path.join(output_dir, filename)
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        
        print(f"Saved post: {title}")

if __name__ == '__main__':
    fetch_and_save_posts()

File: fetch-posts.php
#!/usr/bin/env php
<?php
/**
 * Fetch release posts from the WooCommerce developer blog.
 *
 * Usage: php fetch-posts.php
 *
 * This script fetches posts from the "Release Posts" category on
 * developer.woocommerce.com and saves them as text files.
 */

require_once __DIR__ . '/includes/functions.php';

define('WP_SITE_URL', 'https://developer.woocommerce.com');
define('RELEASE_POSTS_DIR', __DIR__ . '/release-posts');

/**
 * Get date from post and format it as YYYY-MM-DD.
 *
 * @param array $post The post data
 * @return string The formatted date
 */
function get_date_from_post($post) {
    $date = new DateTime($post['date']);
    return $date->format('Y-m-d');
}

/**
 * Fetch and save release posts from WordPress.
 *
 * @return void
 */
function fetch_and_save_posts() {
    $base_url = WP_SITE_URL . '/wp-json/wp/v2';

    // First, get the category ID for 'Release Posts'
    $categories_url = "{$base_url}/categories";
    $categories_response = wordpress_request($categories_url, ['per_page' => 100]);
    $categories = json_decode($categories_response, true);

    if (!is_array($categories)) {
        echo "Error fetching categories\n";
        return;
    }

    // Find the Release Posts category
    $release_posts_category = null;
    foreach ($categories as $cat) {
        if ($cat['name'] === 'Release Posts') {
            $release_posts_category = $cat;
            break;
        }
    }

    if (!$release_posts_category) {
        echo "Could not find 'Release Posts' category\n";
        return;
    }

    // Get posts from the Release Posts category
    $posts_url = "{$base_url}/posts";
    $params = [
        'categories' => $release_posts_category['id'],
        'per_page' => 100,
        'orderby' => 'date',
        'order' => 'desc'
    ];

    $posts_response = wordpress_request($posts_url, $params);
    $posts = json_decode($posts_response, true);

    if (!is_array($posts)) {
        echo "Error fetching posts\n";
        return;
    }

    // Create release-posts directory if it doesn't exist
    ensure_directory(RELEASE_POSTS_DIR);

    // Get list of existing files to avoid duplicates
    $existing_files = [];
    if (is_dir(RELEASE_POSTS_DIR)) {
        $existing_files = scandir(RELEASE_POSTS_DIR);
    }

    // Process each post
    foreach ($posts as $post) {
        // Get the date and title
        $post_date = get_date_from_post($post);
        $title = html_entity_decode($post['title']['rendered'], ENT_QUOTES | ENT_HTML5, 'UTF-8');

        // Skip posts with specific terms in the title
        $skip_terms = ['woocommerce-blocks', 'delayed', 'dot-release'];
        $sanitized_title = sanitize_filename($title);
        $should_skip = false;
        foreach ($skip_terms as $term) {
            if (stripos($sanitized_title, strtolower($term)) !== false) {
                $should_skip = true;
                break;
            }
        }

        if ($should_skip) {
            echo "Skipping filtered post: {$title}\n";
            continue;
        }

        // Create a filename with date prefix
        $filename = "{$post_date}-{$sanitized_title}.txt";

        // Skip if we've already processed this post
        if (in_array($filename, $existing_files)) {
            echo "Skipping already processed post: {$title}\n";
            continue;
        }

        // Prepare the content
        $content = "Title: {$title}\n";
        $content .= "Date: {$post['date']}\n";
        $content .= "Link: {$post['link']}\n";
        $content .= "\nContent:\n";

        // Convert HTML to text (similar to html2text)
        $markdown_content = html_to_text($post['content']['rendered']);
        $content .= $markdown_content;

        // Save to file
        $filepath = RELEASE_POSTS_DIR . '/' . $filename;
        if (file_put_contents($filepath, $content) !== false) {
            echo "Saved post: {$title}\n";
        } else {
            echo "Error saving post: {$title}\n";
        }
    }
}

// Main execution
if (php_sapi_name() === 'cli') {
    fetch_and_save_posts();
}
